File: sample-videochat-webrtc/sample-videochat-webrtc/Components/CallKitManager/CallKitManager.m
//
//  CallKitManager.m
//  sample-videochat-webrtc
//
//  Created by Injoit on 3/12/19.
//  Copyright © 2019 Quickblox. All rights reserved.
//

#import "CallKitManager.h"
#import <CallKit/CallKit.h>
#import "UsersDataSource.h"
#import "Log.h"
#import "AppDelegate.h"

static const NSInteger DefaultMaximumCallsPerCallGroup = 1;
static const NSInteger DefaultMaximumCallGroups = 1;

@interface CallKitManager () <CXProviderDelegate>

@property (assign, nonatomic) BOOL isCallStarted;
@property (strong, nonatomic) CXProvider *provider;
@property (strong, nonatomic) CXCallController *callController;
@property (copy, nonatomic) dispatch_block_t actionCompletionBlock;
@property (copy, nonatomic) CompletionActionBlock onAcceptActionBlock;
@property (weak, nonatomic) QBRTCSession *session;
@property (strong, nonatomic) NSArray<Call *> *calls;
@end

@implementation CallKitManager
// MARK: - Static
+ (CallKitManager *)instance {
    static CallKitManager *instance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        instance = [[CallKitManager alloc] init];
    });
    return instance;
}

+ (CXProviderConfiguration *)configuration {
    NSString *appName = [[[NSBundle mainBundle] infoDictionary] objectForKey:@"CFBundleDisplayName"];
    CXProviderConfiguration *config = [[CXProviderConfiguration alloc] initWithLocalizedName:appName];
    config.supportsVideo = YES;
    config.maximumCallsPerCallGroup = DefaultMaximumCallsPerCallGroup;
    config.maximumCallGroups = DefaultMaximumCallGroups;
    config.supportedHandleTypes = [NSSet setWithObjects:@(CXHandleTypePhoneNumber), nil];
    config.iconTemplateImageData = UIImagePNGRepresentation([UIImage imageNamed:@"CallKitLogo"]);
    config.ringtoneSound = @"ringtone.wav";
    return config;
}

// MARK: - Initialization
- (instancetype)init {
    self = [super init];
    if (self != nil) {
        self.calls = @[];
        CXProviderConfiguration *configuration = [CallKitManager configuration];
        self.provider = [[CXProvider alloc] initWithConfiguration:configuration];
        [self.provider setDelegate:self queue:nil];
        self.callController = [[CXCallController alloc] initWithQueue:dispatch_get_main_queue()];
        self.isCallStarted = NO;
    }
    return self;
}

- (Boolean)isCallDidStarted {
    return self.isCallStarted == YES;
}

- (Call * _Nullable)currentCall {
    Log(@"currentCall self.calls %@",  @(self.calls.count));
    Call *currentCall = (Call *)self.calls.firstObject;
    if (currentCall) {
        return currentCall;
    }
    return nil;
}

- (Boolean)isHasSession {
    return self.session != nil;
}

- (void)setupSession:(QBRTCSession *)session {
    _session = session;
    if ([self.delegate respondsToSelector:@selector(callKitManager:didUpdateSession:)]) {
        [self.delegate callKitManager:self didUpdateSession:session];
    }
}

- (void)setIsCallStarted:(BOOL)isCallStarted {
    if (_isCallStarted != isCallStarted) {
        _isCallStarted = isCallStarted;
        AppDelegate *appDelegate = (AppDelegate*)[UIApplication.sharedApplication delegate];
        appDelegate.isCalling = _isCallStarted;
    }
}

// MARK: - Call management
- (void)startCallWithUserIDs:(NSArray *)userIDs session:(QBRTCSession *)session uuid:(NSUUID *)uuid {
    _session = session;
    
    self.calls = @[];
    Call *outgoingCall = [[Call alloc] initWithUUID:uuid sessionID:session.ID status:CallStatusActive];
    self.calls = @[outgoingCall];
    
    NSString *contactIdentifier = nil;
    CXHandle *handle = [self handleForUserIDs:userIDs];
    CXStartCallAction *action = [[CXStartCallAction alloc] initWithCallUUID:uuid handle:handle];
    action.contactIdentifier = contactIdentifier;
    
    CXTransaction *transaction = [[CXTransaction alloc] initWithAction:action];
    [self requestTransaction:transaction completion:^(__unused BOOL succeed) {
        CXCallUpdate *update = [[CXCallUpdate alloc] init];
        update.remoteHandle = handle;
        update.localizedCallerName = contactIdentifier;
        update.supportsHolding = NO;
        update.supportsGrouping = NO;
        update.supportsUngrouping = NO;
        update.supportsDTMF = NO;
        update.hasVideo = session.conferenceType == QBRTCConferenceTypeVideo;
        
        [self.provider reportCallWithUUID:uuid updated:update];
    }];
}

- (void)endCallWithUUID:(NSUUID *)uuid completion:(dispatch_block_t)completion {
    Call *currentCall = (Call *)self.calls.firstObject;
    if (!currentCall) {
        return;
    }
    
    CXEndCallAction *action = [[CXEndCallAction alloc] initWithCallUUID:currentCall.uuid];
    CXTransaction *transaction = [[CXTransaction alloc] initWithAction:action];
    self.calls = @[];
    
    dispatchOnMainThread(^{
        [self requestTransaction:transaction completion:nil];
    });
    
    if (completion != nil) {
        _actionCompletionBlock = completion;
    }
}

- (void)reportIncomingCallWithUserIDs:(NSArray *)userIDs
                        outCallerName:(NSString *)callerName
                              session:(QBRTCSession * _Nullable)session
                            sessionID:(NSString *)sessionID
                sessionConferenceType:(QBRTCConferenceType)sessionConferenceType
                                 uuid:(NSUUID *)uuid
                       onAcceptAction:(CompletionActionBlock)onAcceptAction
                           completion:(void (^)(BOOL))completion {
    
    _session = session;
    _onAcceptActionBlock = onAcceptAction;
    
    self.calls = @[];
    Call *incomingCall = [[Call alloc] initWithUUID:uuid sessionID:session.ID status:CallStatusInvite];
    self.calls = @[incomingCall];
    
    CXCallUpdate *update = [[CXCallUpdate alloc] init];
    update.localizedCallerName = callerName;
    update.supportsHolding = NO;
    update.supportsGrouping = NO;
    update.supportsUngrouping = NO;
    update.supportsDTMF = NO;
    update.hasVideo = sessionConferenceType == QBRTCConferenceTypeVideo;
    
    if (self.session) {
        QBRTCAudioSession *audioSession = [QBRTCAudioSession instance];
        audioSession.useManualAudio = YES;
        // disabling audio unit for local mic recording in recorder to enable it later
        session.recorder.localAudioEnabled = NO;
        if (!audioSession.isInitialized) {
            [audioSession initializeWithConfigurationBlock:^(QBRTCAudioSessionConfiguration *configuration) {
                // adding blutetooth support
                configuration.categoryOptions |= AVAudioSessionCategoryOptionAllowBluetooth;
                configuration.categoryOptions |= AVAudioSessionCategoryOptionAllowBluetoothA2DP;
                
                // adding airplay support
                configuration.categoryOptions |= AVAudioSessionCategoryOptionAllowAirPlay;
                if (self.session.conferenceType == QBRTCConferenceTypeVideo) {
                    // setting mode to video chat to enable airplay audio and speaker only
                    configuration.mode = AVAudioSessionModeVideoChat;
                } else if (self.session.conferenceType == QBRTCConferenceTypeAudio) {
                    // setting mode to video chat to enable airplay audio and speaker only
                    configuration.mode = AVAudioSessionModeVoiceChat;
                }
            }];
        }
    }
    
    [_provider reportNewIncomingCallWithUUID:uuid update:update completion:^(NSError * _Nullable error) {
        BOOL silent = ([error.domain isEqualToString:CXErrorDomainIncomingCall] && error.code == CXErrorCodeIncomingCallErrorFilteredByDoNotDisturb);
        dispatchOnMainThread(^{
            if (completion != nil) {
                completion(silent);
            }
        });
    }];
}

- (void)updateIncomingCallWithUserIDs:(NSArray *)userIDs
                        outCallerName:(NSString *)callerName
                              session:(QBRTCSession *)session {
    Call *incomingCall = (Call *)self.calls.firstObject;
    if (!incomingCall) {
        return;
    }
    
    if (!self.session) {
        _session = session;
    }
    
    CXCallUpdate *update = [[CXCallUpdate alloc] init];
    update.localizedCallerName = callerName;
    update.supportsHolding = NO;
    update.supportsGrouping = NO;
    update.supportsUngrouping = NO;
    update.supportsDTMF = NO;
    update.hasVideo = session.conferenceType == QBRTCConferenceTypeVideo;
    
    QBRTCAudioSession *audioSession = [QBRTCAudioSession instance];
    audioSession.useManualAudio = YES;
    // disabling audio unit for local mic recording in recorder to enable it later
    session.recorder.localAudioEnabled = NO;
    if (!audioSession.isInitialized) {
        [audioSession initializeWithConfigurationBlock:^(QBRTCAudioSessionConfiguration *configuration) {
            // adding blutetooth support
            configuration.categoryOptions |= AVAudioSessionCategoryOptionAllowBluetooth;
            configuration.categoryOptions |= AVAudioSessionCategoryOptionAllowBluetoothA2DP;
            
            // adding airplay support
            configuration.categoryOptions |= AVAudioSessionCategoryOptionAllowAirPlay;
            if (self.session.conferenceType == QBRTCConferenceTypeVideo) {
                // setting mode to video chat to enable airplay audio and speaker only
                configuration.mode = AVAudioSessionModeVideoChat;
            } else if (self.session.conferenceType == QBRTCConferenceTypeAudio) {
                // setting mode to video chat to enable airplay audio and speaker only
                configuration.mode = AVAudioSessionModeVoiceChat;
            }
        }];
    }
    
    [_provider reportCallWithUUID:incomingCall.uuid updated:update];
}

- (void)updateCallWithUUID:(NSUUID *)uuid connectingAtDate:(NSDate *)date {
    [_provider reportOutgoingCallWithUUID:uuid startedConnectingAtDate:date];
}

- (void)updateCallWithUUID:(NSUUID *)uuid connectedAtDate:(NSDate *)date {
    [_provider reportOutgoingCallWithUUID:uuid connectedAtDate:date];
}

// MARK: - CXProviderDelegate protocol
- (void)providerDidReset:(CXProvider *)__unused provider {
}

- (void)provider:(CXProvider *)__unused provider performStartCallAction:(CXStartCallAction *)action {
    if (_session == nil) {
        [action fail];
        return;
    }
    __weak __typeof(self)weakSelf = self;
    dispatchOnMainThread(^{
        [weakSelf.session startCall:nil];
        weakSelf.isCallStarted = YES;
        [action fulfill];
    });
}

- (void)provider:(CXProvider *)__unused provider performAnswerCallAction:(CXAnswerCallAction *)action {
    Call *incomingCall = (Call *)self.calls.firstObject;
    if (![incomingCall.uuid isEqual:action.callUUID] || incomingCall.status != CallStatusInvite) {
        Log(@"performAnswerCallAction self.calls %@",  @(self.calls.count));
        [action fail];
        return;
    }
    incomingCall.status = CallStatusActive;
    if (self.session) {
        if ([UIDevice currentDevice].systemVersion.integerValue == 10) {
            // Workaround for webrtc on ios 10, because first incoming call does not have audio
            // due to incorrect category: AVAudioSessionCategorySoloAmbient
            // webrtc need AVAudioSessionCategoryPlayAndRecord
            NSError *err = nil;
            if (![[AVAudioSession sharedInstance] setCategory:AVAudioSessionCategoryPlayAndRecord error:&err]) {
                Log(@"[%@] Error setting category for webrtc workaround on ios 10.",  NSStringFromClass([CallKitManager class]));
            }
        }
        [self.session acceptCall:nil];
    }
    
    __weak __typeof(self)weakSelf = self;
    dispatchOnMainThread(^{
        
        weakSelf.isCallStarted = YES;
        [action fulfill];
        
        if (weakSelf.onAcceptActionBlock != nil) {
            weakSelf.onAcceptActionBlock(YES);
            weakSelf.onAcceptActionBlock = nil;
        }
    });
}

- (void)provider:(CXProvider *)__unused provider performEndCallAction:(CXEndCallAction *)action {
    
    if (self.calls.count > 0) {
        Call *call = (Call *)self.calls.firstObject;
        call.status = CallStatusEnded;
        if (![call.uuid isEqual:action.callUUID]) {
            self.calls = @[];
            [action fail];
            return;
        }
    }
    
    if (self.session) {
        QBRTCSession *session = self.session;
        _session = nil;
        
        if (self.isCallStarted == YES) {
            [session hangUp:nil];
            self.isCallStarted = NO;
        } else {
            [session rejectCall:nil];
        }
        self.calls = @[];
        
        [action fulfillWithDateEnded:[NSDate date]];
    }
    
    __weak __typeof(self)weakSelf = self;
    dispatchOnMainThread(^{
        QBRTCAudioSession *audioSession = [QBRTCAudioSession instance];
        audioSession.audioEnabled = NO;
        audioSession.useManualAudio = NO;
        [audioSession deinitialize];
        self.isCallStarted = NO;
        [action fulfillWithDateEnded:[NSDate date]];
        
        if (weakSelf.onAcceptActionBlock != nil) {
            weakSelf.onAcceptActionBlock(NO);
            weakSelf.onAcceptActionBlock = nil;
        }
        
        if (weakSelf.actionCompletionBlock != nil) {
            weakSelf.actionCompletionBlock();
            weakSelf.actionCompletionBlock = nil;
        }
    });
}

- (void)provider:(CXProvider *)__unused provider performSetMutedCallAction:(CXSetMutedCallAction *)action {
    __weak __typeof(self)weakSelf = self;
    dispatchOnMainThread(^{
        if (weakSelf.session) {
            weakSelf.session.localMediaStream.audioTrack.enabled = !action.isMuted;
        }
        [action fulfill];
        
        if (weakSelf.onMicrophoneMuteAction != nil) {
            weakSelf.onMicrophoneMuteAction();
        }
    });
}

- (void)provider:(CXProvider *)__unused provider didActivateAudioSession:(AVAudioSession *)audioSession {
    Log(@"[%@] Activated audio session.",  NSStringFromClass([CallKitManager class]));
    QBRTCAudioSession *rtcAudioSession = [QBRTCAudioSession instance];
    [rtcAudioSession audioSessionDidActivate:audioSession];
    // enabling audio now
    rtcAudioSession.audioEnabled = YES;
}

- (void)provider:(CXProvider *)provider didDeactivateAudioSession:(AVAudioSession *)audioSession {
    Log(@"[%@] Dectivated audio session.",  NSStringFromClass([CallKitManager class]));
    [[QBRTCAudioSession instance] audioSessionDidDeactivate:audioSession];
    // deinitializing audio session after iOS deactivated it for us
    QBRTCAudioSession *rtcAudioSession = [QBRTCAudioSession instance];
    if (rtcAudioSession.isInitialized) {
        Log(@"[%@] Deinitializing session in CallKit callback.",  NSStringFromClass([CallKitManager class]));
        [rtcAudioSession deinitialize];
    }
}

// MARK: - Helpers

- (CXHandle *)handleForUserIDs:(NSArray *)userIDs {
    if (userIDs.count == 1) {
        QBUUser *user = [self.usersDatasource userWithID:[[userIDs firstObject] integerValue]];
        if (user.phone.length > 0) {
            return [[CXHandle alloc] initWithType:CXHandleTypeGeneric value:user.phone];
        }
    }
    
    return [[CXHandle alloc] initWithType:CXHandleTypePhoneNumber value:[userIDs componentsJoinedByString:@", "]];
}

static inline void dispatchOnMainThread(dispatch_block_t block) {
    if ([NSThread isMainThread]) {
        block();
    }
    else {
        dispatch_async(dispatch_get_main_queue(), block);
    }
}

- (void)requestTransaction:(CXTransaction *)transaction completion:(void (^)(BOOL))completion {
    [_callController requestTransaction:transaction completion:^(NSError *error) {
        if (error != nil) {
            Log(@"[%@] Error: %@",  NSStringFromClass([CallKitManager class]), error);
        }
        if (completion) {
            completion(error == nil);
        }
    }];
}

@end

File: sample-videochat-webrtc-swift/sample-videochat-webrtc-swift/Components/CallKitManager/CallKitManager.swift
//
//  CallKitManager.swift
//  sample-videochat-webrtc-swift
//
//  Created by Injoit on 12/10/18.
//  Copyright © 2018 QuickBlox. All rights reserved.
//

import UIKit
import QuickbloxWebRTC
import Quickblox
import CallKit

struct CallKitManagerConstant {
    static let maximumCallsPerCallGroup: Int = 1
    static let maximumCallGroups: Int = 1
}

enum CallStatus: UInt {
    case none = 0
    case invite
    case active
    case ended
}

struct Call {
    var uuid: UUID
    var sessionID: String?
    var status: CallStatus
}

typealias CompletionBlock = (() -> Void)
typealias CompletionActionBlock = ((Bool) -> Void)
typealias AudioSessionInitializeBlock = ((QBRTCSession) -> Void)

protocol CallKitManagerDelegate: class {
    func callKitManager(_ callKitManager: CallKitManager, didUpdateSession session: QBRTCSession)
}
/**
 CallKitManager class interface.
 Used as manager of Apple CallKit.
 */
class CallKitManager: NSObject {
    
    //MARK: - Properties
    weak var delegate: CallKitManagerDelegate?
    
    private var callStarted = false {
        didSet {
            if let appDelegate = UIApplication.shared.delegate as? AppDelegate {
                appDelegate.isCalling = callStarted
            }
        }
    }
    
    /**
     UserDataSource instance to get users information from.
     */
    var usersDatasource: UsersDataSource?
    /**
     Action on microphone mute using CallKit UI.
     */
    var onMicrophoneMuteAction: CompletionBlock?
    private var provider: CXProvider?
    private var callController: CXCallController?
    private var actionCompletionBlock: CompletionBlock?
    private var onAcceptActionBlock: CompletionActionBlock?
    private var audioSessionInitializeBlock: AudioSessionInitializeBlock?
    private var session: QBRTCSession? {
        didSet {
            if let session = session {
                self.delegate?.callKitManager(self, didUpdateSession: session)
                if let audioSessionInitializeBlock = self.audioSessionInitializeBlock {
                    audioSessionInitializeBlock(session)
                    self.audioSessionInitializeBlock = nil
                }
            }
        }
    }
    private var calls: [Call] = []
    
    /**
     Class singleton instance.
     */
    static let instance = CallKitManager()
    
    class func configuration() -> CXProviderConfiguration {
        let appName = Bundle.main.infoDictionary?["CFBundleDisplayName"] as? String
        let config = CXProviderConfiguration(localizedName: appName ?? "")
        config.supportsVideo = true
        config.maximumCallsPerCallGroup = CallKitManagerConstant.maximumCallsPerCallGroup
        config.maximumCallGroups = CallKitManagerConstant.maximumCallGroups
        config.supportedHandleTypes = [.phoneNumber]
        if let image = UIImage(named: "CallKitLogo") {
            config.iconTemplateImageData = image.pngData()
        }
        config.ringtoneSound = "ringtone.wav"
        return config
    }
    
    // MARK: - Life Cycle
    override init() {
        super.init()
        
        provider = CXProvider(configuration: CallKitManager.configuration())
        
        if let provider = provider {
            provider.setDelegate(self, queue: nil)
        }
        callController = CXCallController(queue: DispatchQueue.main)
        callStarted = false
    }
    
    func isCallStarted() -> Bool { 
        return callStarted
    }
    
    func isHasSession() -> Bool {
        return self.session != nil
    }
    
    func currentCall() -> Call? {
        if let currentCall = calls.first {
            return currentCall
        }
        return nil
    }
    
    func setupSession(_ session: QBRTCSession) {
        self.session = session
    }
    
    // MARK: - Call management
    /**
     Start Call with user IDs.
     
     @param userIDs user IDs to perform call with
     @param session session instance
     @param uuid call uuid
     
     @discussion Use this to perform outgoing call with specific user ids.
     
     @see QBRTCSession
     */
    func startCall(withUserIDs userIDs: [NSNumber], session: QBRTCSession?, uuid: UUID?) {
        guard let session = session,
            let uuid = uuid else {
                return
        }
        self.session = session
        
        calls = []
        let outgoingCall = Call(uuid: uuid, sessionID: session.id, status: .active)
        calls = [outgoingCall]
        
        let contactIdentifier = ""
        let handle = self.handle(forUserIDs: userIDs)
        let action = CXStartCallAction(call: uuid, handle: handle)
        action.contactIdentifier = contactIdentifier
        let transaction = CXTransaction(action: action)
        request(transaction) { error in
            if error == nil {
                let update = CXCallUpdate()
                update.remoteHandle = handle
                update.localizedCallerName = contactIdentifier
                update.supportsHolding = false
                update.supportsGrouping = false
                update.supportsUngrouping = false
                update.supportsDTMF = false
                update.hasVideo = session.conferenceType == QBRTCConferenceType.video
                if let provider = self.provider {
                    provider.reportCall(with: uuid, updated: update)
                }
            }
        }
    }
    
    /**
     End call with uuid.
     @param uuid uuid of call
     @param completion completion block
     */
    func endCall(with uuid: UUID?, completion: CompletionBlock? = nil) {
        guard let currentCall = calls.first else {
            return
        }
        let  action = CXEndCallAction(call: currentCall.uuid)
        let transaction = CXTransaction(action: action)
        dispatchOnMainThread(block: {
            self.request(transaction, completion: { (error) in
                if let error = error {
                    debugPrint("[CallKitManager] Request transaction error: \(error.localizedDescription)")
                }
            })
        })
        actionCompletionBlock = completion
    }
    
    /**
     Report incoming call with user IDs.
     
     @param userIDs user IDs of incoming call
     @param session session instance
     @param uuid call uuid
     @param onAcceptAction on call accept action
     @param completion completion block
     
     @discussion Use this to show incoming call screen.
     
     @see QBRTCSession
     */
    func reportIncomingCall(withUserIDs userIDs: [NSNumber],
                            outCallerName: String,
                            session: QBRTCSession?,
                            sessionID: String?,
                            sessionConferenceType:QBRTCConferenceType,
                            uuid: UUID,
                            onAcceptAction: @escaping CompletionActionBlock,
                            completion: @escaping (Bool) -> Void) {
        
        calls = []
        if session != nil {
            self.session = session
        }
        
        onAcceptActionBlock = onAcceptAction
        
        let incomingCall = Call(uuid: uuid, sessionID: sessionID, status: .invite)
        calls = [incomingCall]
        
        let update = CXCallUpdate()
        update.localizedCallerName = outCallerName
        update.supportsHolding = false
        update.supportsGrouping = false
        update.supportsUngrouping = false
        update.supportsDTMF = false
        update.hasVideo = true
        update.hasVideo = sessionConferenceType == .video
        
        let audioSession = QBRTCAudioSession.instance()
        audioSession.useManualAudio = true
        // disabling audio unit for local mic recording in recorder to enable it later
        if audioSession.isInitialized == false {
            audioSession.initialize { configuration in
                // adding blutetooth support
                configuration.categoryOptions.insert(.allowBluetooth)
                configuration.categoryOptions.insert(.allowBluetoothA2DP)
                configuration.categoryOptions.insert(.duckOthers)
                // adding airplay support
                configuration.categoryOptions.insert(.allowAirPlay)
                if sessionConferenceType == .video {
                    // setting mode to video chat to enable airplay audio and speaker only
                    configuration.mode = AVAudioSession.Mode.videoChat.rawValue
                } else if sessionConferenceType == .audio {
                    // setting mode to video chat to enable airplay audio and speaker only
                    configuration.mode = AVAudioSession.Mode.voiceChat.rawValue
                }
            }
        }
        
        if let provider = provider {
            provider.reportNewIncomingCall(with: uuid, update: update) { error in
                let errorDomain = (error as NSError?)?.domain
                let errorCode = (error as NSError?)?.code
                let silent = errorDomain == CXErrorDomainIncomingCall &&
                    errorCode == CXErrorCodeIncomingCallError.Code.filteredByDoNotDisturb.rawValue
                self.dispatchOnMainThread(block: {
                    completion(silent)
                })
            }
        }
    }
    
    func updateIncomingCall(withUserIDs userIDs: [NSNumber],
                            outCallerName: String,
                            session: QBRTCSession,
                            uuid: UUID) {
        guard let provider = provider else {
            return
        }
        
        self.session = session
        let audioSession = QBRTCAudioSession.instance()
        
        audioSession.useManualAudio = true
        // disabling audio unit for local mic recording in recorder to enable it later
        session.recorder?.isLocalAudioEnabled = false
        if audioSession.isInitialized == false {
            audioSession.initialize { configuration in
                // adding blutetooth support
                configuration.categoryOptions.insert(.allowBluetooth)
                configuration.categoryOptions.insert(.allowBluetoothA2DP)
                configuration.categoryOptions.insert(.duckOthers)
                // adding airplay support
                configuration.categoryOptions.insert(.allowAirPlay)
                if session.conferenceType == .video {
                    // setting mode to video chat to enable airplay audio and speaker only
                    configuration.mode = AVAudioSession.Mode.videoChat.rawValue
                } else if session.conferenceType == .audio {
                    // setting mode to video chat to enable airplay audio and speaker only
                    configuration.mode = AVAudioSession.Mode.voiceChat.rawValue
                }
            }
        }
        
        let update = CXCallUpdate()
        update.localizedCallerName = outCallerName
        update.supportsHolding = false
        update.supportsGrouping = false
        update.supportsUngrouping = false
        update.supportsDTMF = false
        update.hasVideo = true
        update.hasVideo = session.conferenceType == .video
        
        provider.reportCall(with: uuid, updated: update)
    }
    
    /**
     Update outgoing call with connecting date
     
     @param uuid call uuid
     @param date connecting started date
     */
    func updateCall(with uuid: UUID?, connectingAt date: Date?) {
        if let uuid = uuid,
            let provider = provider {
            provider.reportOutgoingCall(with: uuid, startedConnectingAt: date)
        }
    }
    
    /**
     Update outgoing call with connected date.
     
     @param uuid call uuid
     @param date connected date
     */
    func updateCall(with uuid: UUID?, connectedAt date: Date?) {
        if let uuid = uuid,
            let provider = provider {
            provider.reportOutgoingCall(with: uuid, connectedAt: date)
        }
    }
    
    // MARK: - Helpers
    func handle(forUserIDs userIDs: [NSNumber]) -> CXHandle {
        if userIDs.count == 1 {
            if let userId = userIDs.first,
                let usersDatasource = usersDatasource,
                let user = usersDatasource.user(withID: userId.uintValue),
                user.phone?.isEmpty == false {
                return CXHandle(type: .phoneNumber, value: user.phone ?? "")
            }
        }
        let arrayUserIDs = userIDs.map({$0.stringValue})
        return CXHandle(type: .phoneNumber, value: arrayUserIDs.joined(separator: ", "))
    }
    
    @inline(__always) private func dispatchOnMainThread(block: @escaping () -> Void) {
        if Thread.isMainThread == true {
            block()
        } else {
            DispatchQueue.main.async(execute: block)
        }
    }
    
    func request(_ transaction: CXTransaction, completion: @escaping (Error?) -> Void) {
        callController?.request(transaction) { error in
            if let error = error {
                debugPrint("[CallKitManager] Error: \(error)")
            }
            debugPrint("[CallKitManager] transaction successfully!!!")
            completion(nil)
        }
    }
}

extension CallKitManager: CXProviderDelegate {
    // MARK: - CXProviderDelegate protocol
    func providerDidReset(_ provider: CXProvider){
    }
    
    func provider(_ provider: CXProvider, perform action: CXStartCallAction) {
        guard let session = self.session else {
            action.fail()
            return
        }
        dispatchOnMainThread(block: { [weak self] in
            session.startCall(nil)
            self?.callStarted = true
            action.fulfill()
        })
    }
    
    func provider(_ provider: CXProvider, perform action: CXAnswerCallAction) {
        if let index = calls.firstIndex(where: { $0.uuid == action.callUUID }),
            calls[index].status == .invite {
            calls[index].status = .active
        } else {
            return
        }
        if let session = self.session {
            if (Int(UIDevice.current.systemVersion) == 10) {
                // Workaround for webrtc on ios 10, because first incoming call does not have audio
                // due to incorrect category: AVAudioSessionCategorySoloAmbient
                // webrtc need AVAudioSessionCategoryPlayAndRecord
                try! AVAudioSession.sharedInstance().setCategory(.playAndRecord)
                debugPrint("[CallKitManager] Error setting category for webrtc workaround on ios 10.")
            }
            self.callStarted = true
            session.acceptCall(nil)
        }
        
        dispatchOnMainThread(block: { [weak self] in
            guard let self = self else {
                return
            }
            action.fulfill()
            if let onAcceptActionBlock = self.onAcceptActionBlock {
                onAcceptActionBlock(true)
                self.onAcceptActionBlock = nil
            }
        })
    }
    
    func provider(_ provider: CXProvider, perform action: CXEndCallAction) {
        if let index = calls.firstIndex(where: { $0.uuid == action.callUUID }) {
            calls[index].status = .ended
            if let session = self.session {
                if callStarted == true {
                    session.hangUp(nil)
                    callStarted = false
                } else {
                    session.rejectCall(nil)
                }
                self.session = nil
                calls = []
            }
        } else {
            callStarted = false
            calls = []
            action.fail()
            return
        }
        
        dispatchOnMainThread(block: { [weak self] in
            guard let self = self else {
                return
            }
            let audioSession = QBRTCAudioSession.instance()
            if audioSession.isInitialized == true {
                audioSession.isAudioEnabled = false
                audioSession.useManualAudio = false
                audioSession.deinitialize()
            }
            action.fulfill(withDateEnded: Date())
            
            if let onAcceptActionBlock = self.onAcceptActionBlock {
                onAcceptActionBlock(false)
                self.onAcceptActionBlock = nil
            }
            if let actionCompletionBlock = self.actionCompletionBlock {
                actionCompletionBlock()
                self.actionCompletionBlock = nil
            }
            
            self.audioSessionInitializeBlock = nil
        })
    }
    
    func provider(_ provider: CXProvider, perform action: CXSetMutedCallAction){
        dispatchOnMainThread(block: { [weak self] in
            if let session = self?.session {
                session.localMediaStream.audioTrack.isEnabled = !action.isMuted
            }
            action.fulfill()
            if let onMicrophoneMuteAction = self?.onMicrophoneMuteAction {
                onMicrophoneMuteAction()
            }
        })
    }
    
    func provider(_ provider: CXProvider, didActivate audioSession: AVAudioSession){
        if self.session != nil {
            debugPrint("[CallKitManager] Activated audio session.")
            activateAudioSession(audioSession)
        } else {
            self.audioSessionInitializeBlock = { [weak self] qbSession in
                self?.activateAudioSession(audioSession)
                debugPrint("[CallKitManager] Activated audio session in audioSessionInitializeBlock.")
            }
        }
    }
    
    private func activateAudioSession(_ audioSession: AVAudioSession) {
        let rtcAudioSession = QBRTCAudioSession.instance()
        rtcAudioSession.audioSessionDidActivate(audioSession)
        // enabling audio now
        rtcAudioSession.isAudioEnabled = true
    }
    
    func provider(_ provider: CXProvider, didDeactivate audioSession: AVAudioSession){
        debugPrint("[CallKitManager] Dectivated audio session.")
        QBRTCAudioSession.instance().audioSessionDidDeactivate(audioSession)
        // deinitializing audio session after iOS deactivated it for us
        let rtcAudioSession = QBRTCAudioSession.instance()
        if rtcAudioSession.isInitialized == true {
            debugPrint("Deinitializing session in CallKit callback.")
            rtcAudioSession.deinitialize()
        }
        self.audioSessionInitializeBlock = nil
    }
}
